moveit_core/planning_scene/src/planning_scene.cpp

```diff
@@ -1551,42 +1558,98 @@ bool PlanningScene::processAttachedCollisionObjectMsg(const moveit_msgs::Attache
                         object.link_name.c_str());
       }
-      else
+      // STEP 3.5: For append operations, only update the attached object and don't look in the world
+      else  // if operation == append
       {
-        const robot_state::AttachedBody* ab = kstate_->getAttachedBody(object.object.id);
-        shapes.insert(shapes.end(), ab->getShapes().begin(), ab->getShapes().end());
-        poses.insert(poses.end(), ab->getFixedTransforms().begin(), ab->getFixedTransforms().end());
+        const robot_state::AttachedBody* ab = kstate_->getAttachedBody(object_to_attach.id_);
+        object_to_attach.shapes_.insert(object_to_attach.shapes_.end(), ab->getShapes().begin(), ab->getShapes().end());
+        object_to_attach.shape_poses_.insert(object_to_attach.shape_poses_.end(), ab->getFixedTransforms().begin(),
+                                             ab->getFixedTransforms().end());
         trajectory_msgs::JointTrajectory detach_posture =
             object.detach_posture.joint_names.empty() ? ab->getDetachPosture() : object.detach_posture;
-        std::set<std::string> ab_touch_links = ab->getTouchLinks();
-        kstate_->clearAttachedBody(object.object.id);
+        std::set<std::string> touch_links;
         if (object.touch_links.empty())
-          kstate_->attachBody(object.object.id, shapes, poses, ab_touch_links, object.link_name, detach_posture);
+        {
+          touch_links = ab->getTouchLinks();
+        }
         else
-          kstate_->attachBody(object.object.id, shapes, poses, object.touch_links, object.link_name, detach_posture);
-        ROS_DEBUG_NAMED("planning_scene", "Added shapes to object '%s' attached to link '%s'", object.object.id.c_str(),
-                        object.link_name.c_str());
-      }
+        {
+          touch_links = std::set<std::string>(std::make_move_iterator(object.touch_links.begin()),
+                                              std::make_move_iterator(object.touch_links.end()));
+        }
+        object_to_attach.named_frames_.insert(ab->getNamedTransforms().begin(), ab->getNamedTransforms().end());
 
+        kstate_->clearAttachedBody(object_to_attach.id_);
+        kstate_->attachBody(object_to_attach.id_, object_to_attach.shapes_, object_to_attach.shape_poses_, touch_links,
+                            object.link_name, detach_posture, object_to_attach.named_frames_);
+        ROS_DEBUG_NAMED("planning_scene", "Appended things to object '%s' attached to link '%s'",
+                        object_to_attach.id_.c_str(), object.link_name.c_str());
+      }
       return true;
     }
     else
       ROS_ERROR_NAMED("planning_scene", "Robot state is not compatible with robot model. This could be fatal.");
   }
-  else if (object.object.operation == moveit_msgs::CollisionObject::REMOVE)
+  else if (object.object.operation == moveit_msgs::CollisionObject::REMOVE)  // == DETACH
```

Comment:
I added it as an explanation because for attached objects this really means "detach" instead of "remove from the scene". I would be in favor of making "DETACH" an explicit codeword, but that's for a different PR.